### ISSUE TITLE
<SignIn> Component Not Rendering After Signing Out When Component Has Appearance Props Added

### ISSUE THREAD

**Issue by brettabamonte** [clerk/javascript#1981]
### Preliminary Checks

- [X] I have reviewed the documentation: https://clerk.com/docs

- [X] I have searched for existing issues: https://github.com/clerkinc/javascript/issues

- [X] I have not already reached out to Clerk support via email or Discord (if you have, no need to open an issue here)

- [X] This issue is not a question, general help request, or anything other than a bug report directly related to Clerk. Please ask questions in our Discord community: https://clerk.com/discord.

### Reproduction / Replay Link

https://github.com/brettabamonte/Clerk-Minimal_Reproduction-SignIn_Component/tree/main

### Publishable key

pk_test_Y2hpZWYtc2N1bHBpbi0xMi5jbGVyay5hY2NvdW50cy5kZXYk

### Description

**Steps to reproduce:**

See steps and screenshots in linked git repository README.md (https://github.com/brettabamonte/Clerk-Minimal_Reproduction-SignIn_Component).

**Expected behavior:**
After clicking the <SignOutButton /> the user should be redirected to the sign in page with a <SignIn /> component displaying that has appearance props.

**Actual behavior:**
After clicking the <SignOutButton /> the user is redirected to the sign in page but the <SignIn /> component does not render when it has the appearance prop set. If the appearance prop is set the component displays properly.

### Environment

```Shell
System:
    OS: macOS 14.0
    CPU: (8) arm64 Apple M2
    Memory: 238.50 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.18.1 - /usr/local/bin/node
    npm: 9.8.1 - /usr/local/bin/npm
  Browsers:
    Chrome: 118.0.5993.117
    Safari: 17.0
  npmPackages:
    @clerk/nextjs: ^4.26.1 => 4.26.1 
    @types/node: 20.1.4 => 20.1.4 
    @types/react: 18.2.6 => 18.2.6 
    @types/react-dom: 18.2.4 => 18.2.4 
    autoprefixer: 10.4.14 => 10.4.14 
    classnames: ^2.3.2 => 2.3.2 
    next: ^14.0.0 => 14.0.0 
    postcss: 8.4.23 => 8.4.23 
    react: 18.2.0 => 18.2.0 
    react-dom: 18.2.0 => 18.2.0 
    tailwindcss: 3.3.2 => 3.3.2 
    typescript: 5.0.4 => 5.0.4
```

**Comment by tjl-jg** [comment]
I have an issue that sounds very similar, but with the `<UserButton />`. After signing out, and signing back in, it renders at size 0x0.

Not sure if it's related or not, but the browser console is showing a couple of errors related to signin. The stack trace is to a packed vendors file, so the line numbers probably aren't relevant, but the error is "Uncaught TypeError: Cannot read properties of undefined (reading 'imageUrl')" and the line of code sets avatarUrl with <minified name = l>.userData.imageUrl.

**Comment by IGassmann** [comment]
I can confirm this also happens on `@clerk/nextjs: 4.27.1`/`next: 14.0.2`. However, it happens for us independently of the use of the `appearance` prop

**Comment by LekoArts** [comment]
Hello all!

Just a short update here to show that I/we haven't forgotten this issue 😅 Thanks for the reproduction, I was able to also see the issue. Important bit: This also happens **without** the appearance prop.
After investigating I'm still not sure exactly what is causing it, might be a race condition or an issue with client side re-hydration.

We'll discuss internally a bit how we want to tackle this but I'd be interested in your opinion on this question (@brettabamonte @tjl-jg @IGassmann):

**What is your expected behavior for the `<SignOutButton />` as to where it should navigate?**
Should it stay on the same page? Navigate to the sign in page? Navigate to a custom page one **has** to define?

**Comment by brettabamonte** [comment]
@LekoArts I agree with @tjl-jg. I like the context to be within the component using a property. Having a default destination is nice if the property isn't declared on the component. 90% of the time, I like redirecting users to a default page like the sign in page but there are instances where it'd be nice to redirect them to a custom destination.

Thank you and the team for looking into this!

**Comment by LekoArts** [comment]
Thanks for your input!

Then you can unblock yourself by using the [`signOutCallback`](https://clerk.com/docs/components/unstyled/sign-out-button#properties) prop on the `<SignOutButton />` component:

```tsx
"use client"

import { useRouter } from "next/navigation"
import { SignOutButton } from "@clerk/nextjs";

export const Signout = () => {
  const router = useRouter()

  return (
    <SignOutButton signOutCallback={() => router.push('/your-path')}>
      <button>Sign Out</button>
    </SignOutButton>
  )
}
```

Or if you want to stay on the same page, a hard refresh will do (in the meantime):

```tsx
"use client"

import { SignOutButton } from "@clerk/nextjs";

export const Signout = () => {
  return (
    <SignOutButton signOutCallback={() => window.location.reload()}>
      <button>Sign Out</button>
    </SignOutButton>
  )
}
```

@brettabamonte What your reproduction shows is in some sense "expected" behavior that is described in this thread here. The blank page shows because it shows the unauthenticated state of the `/dashboard` route which will be empty. The behavior you want kicks in because the middleware redirects to the sign-in page because `/dashboard` is not a public route.
With the callback prop above you can reliably do whatever you want after the signout 👍 

We'll still look into the underlying issue and see if we'd need to change something DX-wise.

**Comment by Sengulair** [comment]
I have the same problem as @tjl-jg. Here is a minimal reproduction by the way: https://github.com/Sengulair/clerk-nextjs-app-issue/commit/f7a29db4ce18ac149714b72f91afff578ddeb677
It still appears on the 5.0.1 version. Do we have some workaround for that case?

**Comment by panteliselef** [comment]
@Sengulair mind giving this another try with [5.0.4](https://www.npmjs.com/package/@clerk/nextjs/v/5.0.4)

**Comment by Acro-Sengulair** [comment]
@panteliselef I've updated the reproduction to 5.0.4. Still failing with the same errors

**Comment by Sengulair** [comment]
@panteliselef I've updated the reproduction to 5.0.4. Still failing with the same errors

**Comment by Sengulair** [comment]
@panteliselef Tested with 5.0.6 and seems the problem is resolved.

**Comment by dimkl** [comment]
Closing this as it seems resolved. @brettabamonte If the issue persists after using a version >= 5.0.6, feel free to re-open the issue.

**Comment by clerk-cookie** [comment]
This issue has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
